sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala

```diff
@@ -44,39 +44,78 @@ abstract class Collect extends ImperativeAggregate {
 
   override def dataType: DataType = ArrayType(child.dataType)
 
-  override def supportsPartial: Boolean = false
-
-  override def aggBufferAttributes: Seq[AttributeReference] = Nil
-
-  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
-
-  override def inputAggBufferAttributes: Seq[AttributeReference] = Nil
-
   // Both `CollectList` and `CollectSet` are non-deterministic since their results depend on the
   // actual order of input rows.
   override def deterministic: Boolean = false
 
-  protected[this] val buffer: Growable[Any] with Iterable[Any]
-
-  override def initialize(b: InternalRow): Unit = {
-    buffer.clear()
+  private def generateOutput(results: Iterable[Any]): Any = {
+    if (results.isEmpty) {
+      null
+    } else {
+      new GenericArrayData(results.toArray)
+    }
   }
 
-  override def update(b: InternalRow, input: InternalRow): Unit = {
+  override def update(buffer: T, input: InternalRow): T = {
+    val value = child.eval(input).asInstanceOf[Any]
+
     // Do not allow null values. We follow the semantics of Hive's collect_list/collect_set here.
     // See: org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMkCollectionEvaluator
-    val value = child.eval(input)
     if (value != null) {
       buffer += value
     }
+    buffer
+  }
+
+  override def merge(buffer: T, other: T): T = {
+    buffer ++= other
   }
 
-  override def merge(buffer: InternalRow, input: InternalRow): Unit = {
-    sys.error("Collect cannot be used in partial aggregations.")
+  override def eval(buffer: T): Any = {
+    generateOutput(buffer)
   }
 
-  override def eval(input: InternalRow): Any = {
-    new GenericArrayData(buffer.toArray)
+  protected def _serialize(obj: Iterable[Any]): Array[Byte] = {
+    val buffer = new Array[Byte](4 << 10)  // 4K
```

Comment:
shall we create `new GenericArrayData(buffer.toArray)` and do unsafe projection?